<html>
	<head>
		<title>Morpheus FAQ</title>
		<script type="text/javascript" src="lmcbutton.js"></script>
	</head>
	<body style="font-family:Tahoma; margin-left:auto; margin-right:auto; width:800px">
		<h1>Morpheus</h1>
		<i>hosted by <a href="https://sourceforge.net/">SourceForge</a></i><br />
		<a href="https://sourceforge.net/"><img src="sourceforge.png" alt="SourceForge" width="120" height="90" /></a>
		<h2>Frequently Asked Questions (FAQ)</h2>
		<p>
			<b>Where did the name "Morpheus" come from?</b>
			<br />
			From the movie <a href="http://en.wikipedia.org/wiki/The_Matrix">The Matrix</a>, of course. Greek mythology is a good source for names when you can't come up with anything meaningful.
		</p>
		<p>
			<b>How can I generate mzML data for Morpheus?</b>
			<br />
			The <a href="http://proteowizard.sourceforge.net/tools/msconvert.html">msconvert</a> tool of <a href="http://proteowizard.sourceforge.net/">ProteoWizard</a> is your best bet, although most mzML should work regardless of what generated it. It works with pretty much any vendor data. However if you are using Thermo .raw files or Agilent .d directories, there are versions which support those formats directly which will remove a step of work. There will be small differences in the results because the vendor's charge state assignment and deisotoping routines are used instead of the generic ones in Morpheus.
		</p>
		<p>
			<b>Where can I get FASTA proteome databases?</b>
			<br />
			They are available from many sources but I recommend <a href="http://www.uniprot.org/">UniProt</a>, particularly the <a href="http://www.uniprot.org/taxonomy/complete-proteomes">complete and reference proteome sets</a>. After you filter to the appropriate proteins, click the orange 'Download' button near the top-right of the page. Under FASTA, click 'Download' for either the canonical sequence data or canonical and isoform sequence data. The former will be faster and result in simpler protein groups, but you might miss out on identifying an interesting isoform. I usually err on the side of databases that are too large rather than too small so my suggestion would be canonical and isoform sequence data.
		</p>
		<p>
			<b>What does the 'Create Target&ndash;Decoy Database On The Fly' checkbox do?</b>
			<br />
			Target&ndash;decoy database searching is an empirical way to estimate how many false positives you have identified, enabling calculation of false discovery rate (FDR): false positives / (true positives + false posivites)&mdash;in other words, out of all the identifications you report, how many are incorrect. Target&ndash;decoy FDR is ubiquitious in mass spectrometry&ndash;based proteomics, and Morpheus is designed around the assumption that all users will employ this strategy. There are two ways to do so. The easiest is to supply a normal, target-only FASTA database. Alternatively, you can use an existing concatenated target&ndash;decoy FASTA database, provided the decoy entries contain the text 'DECOY_' somewhere in the identifiers. When you load your FASTA, Morpheus will do a quick survey to determine if it contains decoys. If not, it will automatically check the 'Create Target&ndash;Decoy Database On The Fly' checkbox. If it does contain decoys, it will uncheck it. You should only override this automatic selection if you really know what you're doing. <i>Note: If your FASTA database does contain decoys, but they aren't recognized as such by Morpheus, you will probably end up with a very low number of identifications. This is because it will make decoys out of decoys, which often entails reversing the sequence twice, thus generating decoys that are actually the normal forward target sequence.</i>
		</p>
		<p>
			<b>I get an error/exception that mentions out of memory. How can I fix this?</b>
			<br />
			There are several things you can try, listed in order of increasing difficulty:
			<ol>
				<li>Re-run the program with the 'Minimize Memory Usage' checkbox checked. This will prevent the program from remembering which peptides it has already searched, therefore reducing the memory usage. It will cause it to run slower, however.</li>
				<li>Close other programs open on your computer while you run Morpheus.</li>
				<li>Try searching a smaller database. For example, are you searching nr or all of SwissProt? Limit it to only the organism(s) of interest. Or are you searching UniProt (SwissProt + TrEMBL) for a specific organism? Maybe you only need to search SwissProt (manually annotated proteins).</li>
				<li>Try on a different computer with more memory.</li>
				<li>Add physical memory to your computer. <i>Remember if you're running a 32-bit operating system, 4 GB is the most you can use.</i></li>
			</ol>
		</p>
		<p>
			<b>What does the "Maximum FDR" control?</b>
			<br />
			This dictates the peptide&ndash;spectrum match (PSM), distinct peptide, and protein group numbers shown in the log file. It does not filter the corresponding outputs, which always contain all hits. It is also used as a filter for determining which peptides are used for determining protein identifications.
		</p>
		<p>
			<b>How should I view the PSM, distinct peptide, and protein group TSV outputs?</b>
			<br />
			Microsoft Excel or most other spreadsheet programs. For the confident identifications, I recommend the filtering function to look at only rows with column Q-Value (%) <= 1, and also column Target? = TRUE so you don't bother looking at decoys.
		</p>
		</p>
	</body>
</html>